src/generic/Matrix.jl

```diff
@@ -157,63 +159,44 @@ end
 #
 ###############################################################################
 
+# create a zero matrix
 function (a::MatSpace{T})() where {T <: NCRingElement}
-   R = base_ring(a)
-   entries = Array{T}(undef, a.nrows, a.ncols)
-   for i = 1:a.nrows
-      for j = 1:a.ncols
-         entries[i, j] = zero(R)
-      end
-   end
-   z = MatSpaceElem{T}(R, entries)
-   return z
+   return dense_matrix_type(T)(base_ring(a), nrows(a), ncols(a))
 end
 
-function (a::MatSpace{T})(b::S) where {S <: NCRingElement, T <: NCRingElement}
+# create a matrix with b on the diagonal
+function (a::AbstractAlgebra.Generic.MatSpace)(b::NCRingElement)
+   M = a()  # zero matrix
    R = base_ring(a)
-   entries = Array{T}(undef, a.nrows, a.ncols)
    rb = R(b)
-   for i = 1:a.nrows
-      for j = 1:a.ncols
-         if i != j
-            entries[i, j] = zero(R)
-         else
-            entries[i, j] = rb
-         end
-      end
+   for i in 1:min(nrows(a), ncols(a))
+      M[i, i] = rb
    end
-   z = MatSpaceElem{T}(R, entries)
-   return z
+   return M
 end
 
-function (a::MatSpace{T})(b::Matrix{T}) where T <: NCRingElement
+# convert a Julia matrix
+function (a::MatSpace{T})(b::AbstractMatrix{S}) where {T <: NCRingElement, S}
+   _check_dim(nrows(a), ncols(a), b)
    R = base_ring(a)
-   _check_dim(a.nrows, a.ncols, b)
-   if !isempty(b)
-      R != parent(b[1, 1]) && error("Unable to coerce matrix")
+
+   # minor optimization for MatSpaceElem
+   if S === T && dense_matrix_type(T) === MatSpaceElem{T} && all(x -> R == parent(x), b)
```

Comment:
This `all(x -> R == parent(x), b)` was not there before -- instead we only checked the parent of `b[1,1]` and somehow hoped that's enough... Of course this check makes things a bit slower, but also safer...